pkg/ccl/changefeedccl/kvfeed/kv_feed_test.go

```diff
@@ -645,3 +700,151 @@ func TestCopyFromSourceToDestUntilTableEvent(t *testing.T) {
+				ResolvedTS: hlc.Timestamp{WallTime: int64(ts)},
+			},
+		}
+	}
+
+	for name, tc := range map[string]struct {
+		spans                    []roachpb.Span
+		withFrontierQuantization bool
+		events                   []*kvpb.RangeFeedEvent
+		expectedFrontierEntries  int
+		expectedFrontier         hlc.Timestamp
+	}{
+		"overlapping spans with close ts with quantization": {
+			withFrontierQuantization: true,
+			spans:                    []roachpb.Span{{Key: []byte("a"), EndKey: []byte("d")}},
+			events: []*kvpb.RangeFeedEvent{
+				makeCheckpointEvent([]byte("a"), []byte("b"), 12),
+				makeCheckpointEvent([]byte("b"), []byte("c"), 13),
+				makeCheckpointEvent([]byte("c"), []byte("d"), 14),
+			},
+			expectedFrontierEntries: 1,
+			expectedFrontier:        hlc.Timestamp{WallTime: int64(10)},
+		},
+		"overlapping spans with close ts without quantization": {
+			withFrontierQuantization: false,
+			spans:                    []roachpb.Span{{Key: []byte("a"), EndKey: []byte("d")}},
+			events: []*kvpb.RangeFeedEvent{
+				makeCheckpointEvent([]byte("a"), []byte("b"), 12),
+				makeCheckpointEvent([]byte("b"), []byte("c"), 13),
+				makeCheckpointEvent([]byte("c"), []byte("d"), 14),
+			},
+			expectedFrontierEntries: 3,
+			expectedFrontier:        hlc.Timestamp{WallTime: int64(12)},
+		},
+		"non-overlapping spans with close ts with quantization": {
+			withFrontierQuantization: true,
+			spans:                    []roachpb.Span{{Key: []byte("a"), EndKey: []byte("z")}},
+			events: []*kvpb.RangeFeedEvent{
+				makeCheckpointEvent([]byte("a"), []byte("b"), 12),
+				makeCheckpointEvent([]byte("b"), []byte("c"), 13),
+				makeCheckpointEvent([]byte("c"), []byte("d"), 14),
+				makeCheckpointEvent([]byte("k"), []byte("m"), 12),
+			},
+			expectedFrontierEntries: 4,
+			expectedFrontier:        hlc.Timestamp{WallTime: int64(0)},
+		},
+		"non-overlapping spans with close ts without quantization": {
+			withFrontierQuantization: false,
+			spans:                    []roachpb.Span{{Key: []byte("a"), EndKey: []byte("z")}},
+			events: []*kvpb.RangeFeedEvent{
+				makeCheckpointEvent([]byte("a"), []byte("b"), 12),
+				makeCheckpointEvent([]byte("b"), []byte("c"), 13),
+				makeCheckpointEvent([]byte("c"), []byte("d"), 14),
+				makeCheckpointEvent([]byte("k"), []byte("m"), 12),
+			},
+			expectedFrontierEntries: 6,
+			expectedFrontier:        hlc.Timestamp{WallTime: int64(0)},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			frontier, err := span.MakeFrontier(tc.spans...)
+			require.NoError(t, err)
+			const quantize = time.Duration(10)
+			for _, e := range tc.events {
+				if tc.withFrontierQuantization {
+					e.Checkpoint.ResolvedTS = mockQuantization(e.Checkpoint.ResolvedTS, quantize)
+				}
+				_, err := frontier.Forward(e.Checkpoint.Span, e.Checkpoint.ResolvedTS)
+				require.NoError(t, err)
+			}
+			frontier.Entries(func(sp roachpb.Span, ts hlc.Timestamp) span.OpResult {
+				t.Logf("span: %v, ts: %v\n", sp, ts)
```

Comment:
nit rm

Reply:
I thought it might be worth leaving it in since it was helpful for debugging test failures wdyt?

Reply:
🤷 if you think it will be helpful and it's not too spammy we can leave it in